README.md
<h1 align="center">Добро пожаловать в netschoolapi</h1>

![GitHub tag (latest SemVer)](https://img.shields.io/github/v/tag/nm17/netschoolapi) ![GitHub](https://img.shields.io/github/license/nm17/netschoolapi) [![HitCount](http://hits.dwyl.io/nm17/netschoolapi.svg)](http://hits.dwyl.io/nm17/netschoolapi) [![Downloads](https://pepy.tech/badge/netschoolapi)](https://pepy.tech/project/netschoolapi) [![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/3675/badge)](https://bestpractices.coreinfrastructure.org/projects/3675)

## :dancing_men: Авторы

:bust_in_silhouette: [**Даниил Николаев**](https://github.com/nm17)

:business_suit_levitating: [**Горланов Иван**](https://github.com/igorlanov)

:mage_man: [**Елисей Власенко**](https://github.com/homus32)

## :handshake: Вклады в проект

Вклады, вопросы и пожелания можно высказать в [чате](https://t.me/netschoolapi) проекта в Телеграме
или во вкладке [Issues](https://github.com/nm17/netschoolapi/issues).

## :balance_scale: Лицензия

Copyright © 2020 Даниил Николаев, Иван Горланов, Елисей Власенко.
Прокект находится под лицензией [MIT](LICENSE).


CHANGELOG.md
# [](https://github.com/nm17/netschoolapi/compare/v2.0.0...v) (2020-09-11)



# [2.0.0](https://github.com/nm17/netschoolapi/compare/v1.5.1...v2.0.0) (2020-09-11)


### Bug Fixes

* make lesson a dataclass ([23b20a2](https://github.com/nm17/netschoolapi/commit/23b20a26840bb0c290a16cd9cfaf9d252883fb2a))
* remove -2 on user_id ([8e56233](https://github.com/nm17/netschoolapi/commit/8e56233ea7a23703f3d49747634f2e40bc1bfab8))
* rename function in login() ([d00ccb6](https://github.com/nm17/netschoolapi/commit/d00ccb671adfb10396cded69895de1365c6bea07))


### Features

* bump version 1.5.1 → 2.0.0 ([5236569](https://github.com/nm17/netschoolapi/commit/523656909e0c50eaff5c13376941236117225080))
* support loginform selection ([4835aab](https://github.com/nm17/netschoolapi/commit/4835aab97ea0b3166a396a34546f52a9ac7eed09))



## [1.5.1](https://github.com/nm17/netschoolapi/compare/v1.5.0...v1.5.1) (2020-09-02)


### Bug Fixes

* make pandas a seperate dependency ([4f5dc4e](https://github.com/nm17/netschoolapi/commit/4f5dc4ece1bec24b89d5fd667fa89d5aa12fca48))


### Features

* bump version 1.5.0 → 1.5.1 ([f250b6d](https://github.com/nm17/netschoolapi/commit/f250b6d4c73752ba70937f93e96a6851fc64ca9b))
* delete docs ([12dc41d](https://github.com/nm17/netschoolapi/commit/12dc41d2c4283d099ebdf9c1e9957af31a7f7ec5))



# [1.5.0](https://github.com/nm17/netschoolapi/compare/v1.4.4...v1.5.0) (2020-03-14)


### Features

* add announcements support ([5b54a60](https://github.com/nm17/netschoolapi/commit/5b54a6071006ce7673af1c65c257018a29db12b4))
* bump version 1.4.4 → 1.5.0 ([538b836](https://github.com/nm17/netschoolapi/commit/538b83615430eb09570867e0e0ca6a3596f919bf))



## [1.4.4](https://github.com/nm17/netschoolapi/compare/v1.4.3...v1.4.4) (2020-03-14)


### Bug Fixes

* russian passwords are now encoded properly ([ae6e4df](https://github.com/nm17/netschoolapi/commit/ae6e4df0db8215fea348ecc596705370307dcbde))


### Features

* bump version 1.4.3 → 1.4.4 ([2bd6948](https://github.com/nm17/netschoolapi/commit/2bd6948773b90f327aa974fe0dff539e6b3c8a36))



## [1.4.3](https://github.com/nm17/netschoolapi/compare/v1.4.2...v1.4.3) (2020-03-11)


### Bug Fixes

* new sgo update, new values ([4533b72](https://github.com/nm17/netschoolapi/commit/4533b728b36517c5c2026349fceab8c6301ddf74))


### Features

* bump version 1.4.2 → 1.4.3 ([e599c50](https://github.com/nm17/netschoolapi/commit/e599c5017fe187fb3fe82b41294f097c3d4b7235))



## [1.4.2](https://github.com/nm17/netschoolapi/compare/v1.4.1...v1.4.2) (2020-02-23)


### Bug Fixes

* return actual marks ([4c65a6b](https://github.com/nm17/netschoolapi/commit/4c65a6b351f3ebe06f5e66e3fa0735d7d468816b))


### Features

* bump version 1.4.1 → 1.4.2 ([8a46756](https://github.com/nm17/netschoolapi/commit/8a4675697de63d88288428f9c9b8e279dbc79e82))



## [1.4.1](https://github.com/nm17/netschoolapi/compare/v1.4.0...v1.4.1) (2020-02-21)


### Features

* bump version 1.4.0 → 1.4.1 ([5603af3](https://github.com/nm17/netschoolapi/commit/5603af320866b5c9307fb8cc1597ea279f014ff2))



# [1.4.0](https://github.com/nm17/netschoolapi/compare/v1.3.0...v1.4.0) (2020-02-21)


### Features

* add .bumpversion.cfg ([cfcdca3](https://github.com/nm17/netschoolapi/commit/cfcdca3d181734314a8c2a1803701a032233b58c))
* add ability to get results as pandas data frames ([6c7feed](https://github.com/nm17/netschoolapi/commit/6c7feedf10379b013fbab08c7d228cf83996edab))
* bump version 1.3.0 → 1.4.0 ([1a9e3cb](https://github.com/nm17/netschoolapi/commit/1a9e3cb36999e5c5c2649de04c5786cafa15c5b8))



# [1.3.0](https://github.com/nm17/netschoolapi/compare/v1.2.0...v1.3.0) (2020-02-16)


### Bug Fixes

* get all resp data ([a3fdf09](https://github.com/nm17/netschoolapi/commit/a3fdf092ec153070cff2371982680ab641c3dcfa))
* use httpx async client ([4484da4](https://github.com/nm17/netschoolapi/commit/4484da4033f8e4da26b63f356502db620606e55a))


### Features

* add CHANGELOG.md ([e846b09](https://github.com/nm17/netschoolapi/commit/e846b09e6983a74f6a594933c119ee672e9c49f6))
* add city and oo selection ([82f72c4](https://github.com/nm17/netschoolapi/commit/82f72c48b5b8cf73ef80252a5b295db49cdf36bb))
* add hits counter ([abced4c](https://github.com/nm17/netschoolapi/commit/abced4cd9512e8612fb2b2fe4dd4d03a46b489f1))
* add logout ([ff96c48](https://github.com/nm17/netschoolapi/commit/ff96c48e2a98895fa59d14530a49418a2ef61682))


netschoolapi/data.py
from dataclasses import dataclass, field
from datetime import datetime, date, time
from typing import Optional, List

from dataclasses_json import dataclass_json, config, LetterCase


# Некоторые ключи пропущены, и это не ошибка.
ASSIGNMENT_TYPES = {
    1: "Практическая работа",
    2: "Тематическая работа",
    3: "Домашнее задание",
    4: "Контрольная работа",
    5: "Самостоятельная работа",
    6: "Лабораторная работа",
    7: "Проект",
    8: "Диктант",
    9: "Реферат",
    10: "Ответ на уроке",
    11: "Сочинение",
    12: "Изложение",
    13: "Зачёт",
    14: "Тестирование",
    16: "Диагностическая контрольная работа",
    17: "Диагностическая работа",
    18: "Контрольное списывание",
    21: "Работа на уроке",
    22: "Работа в тетради",
    23: "Ведение рабочей тетради",
    24: "Доклад/Презентация",
    25: "Проверочная работа",
    26: "Чтение наизусть",
    27: "Пересказ текста",
    29: "Предметный диктант",
    31: "Дифференцированный зачет",
    32: "Работа с картами",
    33: "Экзамен",
    34: "Изложение с элементами сочинения",
    35: "Контроль аудирования",
    36: "Контроль грамматики",
    37: "Контроль лексеки",
    38: "Устный развернутый ответ",
}

DATETIME_DECODER = datetime.fromisoformat
DATE_DECODER = lambda d: (datetime.fromisoformat(d).date())
TIME_DECODER = time.fromisoformat


@dataclass_json(letter_case=LetterCase.CAMEL)
@dataclass(frozen=True)
class Attachment:
    """Приложение. В терминах СГО: прикреплённый файл к домашнему заданию."""

    id: int
    file_name: str
    name: Optional[str] = None


@dataclass_json(letter_case=LetterCase.CAMEL)
@dataclass(frozen=True)
class DetailedAssignment:
    id: int
    assignment_name: str
    activity_name: Optional[str]
    problem_name: Optional[str]
    subject: str = field(metadata=config(
        decoder=lambda s: s["name"], field_name="subjectGroup",
    ))
    is_deleted: bool
    day: date = field(metadata=config(decoder=DATE_DECODER, field_name="date"))
    description: str
    teacher: str = field(metadata=config(decoder=lambda t: t["name"]))


@dataclass_json()
@dataclass(frozen=True)
class Assignment:
    """То же самое, что строчка в дневнике."""

    id: int
    type: str = field(metadata=config(
        decoder=lambda at: ASSIGNMENT_TYPES[at], field_name="typeId",
    ))
    name: str = field(metadata=config(field_name="assignmentName"))
    mark: Optional[int] = field(metadata=config(
        decoder=lambda m: m["mark"] if m else None,
    ))
    deadline: Optional[date] = field(metadata=config(
        decoder=DATE_DECODER, field_name="dueDate",
    ))


@dataclass_json(letter_case=LetterCase.CAMEL)
@dataclass(frozen=True)
class Lesson:

    subject: str = field(metadata=config(field_name="subjectName"))
    day: date = field(metadata=config(decoder=DATE_DECODER))
    starts_at: time = field(metadata=config(decoder=TIME_DECODER, field_name="startTime"))
    ends_at: time = field(metadata=config(decoder=TIME_DECODER, field_name="endTime"))
    number: int
    # relay: int  # Даже не знаю, зачем он может быть нужен
    room: Optional[int]
    assignments: Optional[List[Assignment]] = field(default_factory=list)


@dataclass_json(letter_case=LetterCase.CAMEL)
@dataclass(frozen=True)
class WeekDay:

    day: date = field(metadata=config(decoder=DATE_DECODER, field_name="date"))
    lessons: List[Lesson]


@dataclass_json(letter_case=LetterCase.CAMEL)
@dataclass(frozen=True)
class Diary:

    week_start: date = field(metadata=config(decoder=DATE_DECODER))
    week_end: date = field(metadata=config(decoder=DATE_DECODER))
    schedule: List[WeekDay] = field(metadata=config(field_name="weekDays"))


@dataclass_json(letter_case=LetterCase.CAMEL)
@dataclass(frozen=True)
class Announcement:
    id: int
    name: str
    description: str
    posted_at: date = field(metadata=config(decoder=DATE_DECODER, field_name="postDate"))
    delete_date: Optional[date] = field(metadata=config(decoder=DATE_DECODER))
    # author: User
    # recipientInfo: Optional[str]
    attachments: Optional[List[Attachment]] = field(default_factory=list)
